include/swift/LTO/LTO.h

```diff
@@ -0,0 +1,42 @@
+//===--- LTO.cpp - Swift LTO ----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include <memory>
+#include <functional>
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace swift {
+
+class ASTContext;
+
+namespace lto {
+
+class LTOPipelineImpl;
+class InputFile;
+
+using AddStreamFn = std::function<std::unique_ptr<llvm::raw_ostream>(llvm::StringRef ModuleName)>;
+
+class LTOPipeline {
+public:
+  LTOPipeline();
+  ~LTOPipeline();
+  bool addModule(std::unique_ptr<llvm::MemoryBuffer> Buffer);
+  bool emitLLVMModules(AddStreamFn AddStream);
+
+private:
+  LTOPipelineImpl *Impl;
```

Comment:
Do you expect that the linker plugin will reside in the lld repository?  Won't that diverge really quickly?

Reply:
We could avoid the pimpl pattern, and just vend the extension from Swift.  This would be part of the toolchain side of the build which already depends on the LLVM project, which after the monorepo includes lld.  So, if the consumer builds lld, we can build the plugin and avoid this.

Reply:
I haven't thought about this at all, but we should make sure that ld64 can take advantage of this.

Reply:
I see. lld will be shipped as a part of Swift toolchain, so compnerd's approach makes sense for lld. I think ld64 can take the same approach.